File: core/NextPermutation/NextPermutation.py
from typing import List


def nextPermutation(nums: List[int]) -> None:
    length = len(nums)
    left, right = length - 2, length - 1
    while left >= 0 and nums[left] >= nums[left + 1]:
        left -= 1

    if left >= 0:
        while nums[right] <= nums[left]:
            right -= 1
        nums[left], nums[right] = nums[right], nums[left]
    left += 1
    right = length - 1

    while left < right:
        nums[left], nums[right] = nums[right], nums[left]
        left += 1
        right -= 1

File: core/NextPermutation/NextPermutation.java
package com.dbc;

public class NextPermutation {
    private void swap(int[] nums, int a, int b) {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
    public void nextPermutation(int[] nums) {
        int left = nums.length - 2, right = nums.length - 1;
        while (left >= 0 && nums[left] >= nums[left + 1]) left--;
        if (left >= 0) {
            while (nums[left] >= nums[right]) right--;
            swap(nums, left, right);
        }
        left++;
        right = nums.length - 1;
        while (left < right) {
            swap(nums, left, right);
            left++;
            right--;
        }
    }
}
